models/src/node/drain.rs

```diff
@@ -0,0 +1,364 @@
+// `kubectl drain` similarly waits 5 seconds between eviction attempts.
+const EVICTION_RETRY_INTERVAL: Duration = Duration::from_secs(5);
+
+// After evictions are created, we wait for the Pods to be deleted by Kubernetes.
+// These constants define the poll interval for checking the Pods, as well as the max amount of time to wait.
+const DELETION_CHECK_INTERVAL: Duration = Duration::from_secs(5);
+// `kubectl drain` by default will wait "forever" for an eviction to complete. We follow suit.
+const DELETION_TIMEOUT: Duration = Duration::from_secs(u64::MAX);
+
+// Some errors while attempting evictions result in retries with exponential backoff.
+// These values configure how long to delay between tries.
+// We should be tenacious in attempting retries, as some workloads are sensitive to being suddenly interrupted.
+const RETRY_BASE_DELAY: Duration = Duration::from_millis(100);
+const RETRY_MAX_DELAY: Duration = Duration::from_secs(20);
+const NUM_RETRIES: usize = 10;
+
+/// Wrapper struct to provide retry configurations for evictions.
+struct RetryStrategy {}
+impl RetryStrategy {
+    fn retry_strategy() -> impl Iterator<Item = Duration> {
+        ExponentialBackoff::from_millis(RETRY_BASE_DELAY.as_millis() as u64)
+            .max_delay(RETRY_MAX_DELAY)
+            .map(jitter)
+            .take(NUM_RETRIES)
+    }
+}
+
+impl tokio_retry::Condition<error::EvictionError> for RetryStrategy {
+    fn should_retry(&mut self, error: &error::EvictionError) -> bool {
+        error.should_retry()
+    }
+}
+
+/// Drains a node of all pods.
+///
+/// The Kubernetes API does not provide an implementation of drain. You must use Pod deletion or the Eviction API
+/// to remove all Pods from a given Node. We opt to use Evictions in order to respect Pod Disruption Budgets.
+///
+/// The implementation of `kubectl drain` can be used as inspiration. While this implementation will halt under
+/// certain special conditions, we have slightly different behavior:
+///
+/// Kubectl by default will not evict nodes under some criteria without further instruction.
+/// By default, we ignore:
+/// * DaemonSet Pods - The DaemonSet controller will not respect node cordons, so we don't battle it.
+/// * Mirror Pods - These are static and cannot be controlled.
+///
+/// Otherwise, we evict pods that kubectl gives special care:
+/// - Pods with local storage that will be deleted when drained (emptyDir).
+/// - Unreplicated pods (Pods without a controller.)
+///
+/// PodDisruptionBudgets can be used to protect these workloads from being unduly interrupted.
+#[instrument(skip(k8s_client), err)]
+pub(crate) async fn drain_node(
+    k8s_client: &kube::Client,
+    node_name: &str,
+) -> Result<(), error::DrainError> {
+    let target_pods = find_target_pods(k8s_client, node_name).await?;
+
+    // Perform the eviction for each pod simultaneously.
+    stream::iter(target_pods)
```

Comment:
iter seems [deprecated](https://docs.rs/futures/0.1.31/futures/stream/fn.iter.html), maybe use [iter_result](https://docs.rs/futures/0.2.0/futures/stream/fn.iter_result.html)?

Reply:
I think that might only be true in older versions of `futures`? Looks like in `0.3` anyways that the [deprecation notice is gone](https://docs.rs/futures/0.3.19/futures/stream/fn.iter.html). I think it's because they changed it to not explicitly deal with `Result` types, but instead to act on any arbitrary iterator.